### ISSUE TITLE
Stuck in "Unexpected uncompressed size of file"

### ISSUE THREAD

**Issue by bobrik** [yandex/ClickHouse#1471]
We have the following happening in a loop on a few machines:

```
2017.11.09 23:20:54.486077 [ 9 ] <Error> r0..inner.requests_1m (StorageReplicatedMergeTree): Code: 228, e.displayText() = DB::Exception: Unexpected uncompressed size of file browserMap%2EpageViews.bin in data part, e.what() = DB::Exception. Data after merge is not byte-identical to data on another replicas. There could be several reasons: 1. Using newer version of compression library after server update. 2. Using another compression method. 3. Non-deterministic compression algorithm (highly unlikely). 4. Non-deterministic merge algorithm due to logical error in code. 5. Data corruption in memory due to bug in code. 6. Data corruption in memory due to hardware issue. 7. Manual modification of source data after server startup. 8. Manual modification of checksums stored in ZooKeeper. We will download merged part from replica to force byte-identical result.
2017.11.09 23:20:54.645048 [ 9 ] <Error> DB::StorageReplicatedMergeTree::queueTask()::<lambda(DB::StorageReplicatedMergeTree::LogEntryPtr&)>: Code: 228, e.displayText() = DB::Exception: Unexpected uncompressed size of file browserMap%2EpageViews.bin in data part, e.what() = DB::Exception, Stack trace:
0. /usr/bin/clickhouse-server(StackTrace::StackTrace()+0x16) [0x345b036]
1. /usr/bin/clickhouse-server(DB::Exception::Exception(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, int)+0x1f) [0x14e06ef]
2. /usr/bin/clickhouse-server(DB::MergeTreeDataPartChecksum::checkEqual(DB::MergeTreeDataPartChecksum const&, bool, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) const+0x26c) [0x35e8edc]
3. /usr/bin/clickhouse-server(DB::MergeTreeDataPartChecksums::checkEqual(DB::MergeTreeDataPartChecksums const&, bool) const+0x14a) [0x35e943a]
4. /usr/bin/clickhouse-server(DB::StorageReplicatedMergeTree::checkPartAndAddToZooKeeper(std::shared_ptr<DB::MergeTreeDataPart const> const&, std::vector<std::unique_ptr<zkutil::Op, std::default_delete<zkutil::Op> >, std::allocator<std::unique_ptr<zkutil::Op, std::default_delete<zkutil::Op> > > >&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >)+0x864) [0x355fd04]
5. /usr/bin/clickhouse-server(DB::StorageReplicatedMergeTree::fetchPart(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, bool, unsigned long)+0x4d7) [0x356b7f7]
6. /usr/bin/clickhouse-server(DB::StorageReplicatedMergeTree::executeLogEntry(DB::ReplicatedMergeTreeLogEntry const&)+0x1d8a) [0x356e21a]
7. /usr/bin/clickhouse-server() [0x35702cd]
8. /usr/bin/clickhouse-server(DB::ReplicatedMergeTreeQueue::processEntry(std::function<std::shared_ptr<zkutil::ZooKeeper> ()>, std::shared_ptr<DB::ReplicatedMergeTreeLogEntry>&, std::function<bool (std::shared_ptr<DB::ReplicatedMergeTreeLogEntry>&)>)+0x4a) [0x365a0ba]
9. /usr/bin/clickhouse-server(DB::StorageReplicatedMergeTree::queueTask()+0x150) [0x355b0c0]
10. /usr/bin/clickhouse-server(DB::BackgroundProcessingPool::threadFunction()+0x43e) [0x35b05be]
11. /usr/bin/clickhouse-server() [0x4334eff]
12. /lib/x86_64-linux-gnu/libpthread.so.0(+0x7494) [0x7f4757ac5494]
13. /lib/x86_64-linux-gnu/libc.so.6(clone+0x3f) [0x7f47570f7aff]
```

For this particular machine I found the queue item:

```
[zk: clickhouse1.zookeeper.cfdata.org(CONNECTED) 53] get /clickhouse/foo/tables/7/requests_1m/replicas/mych12/queue/queue-0000822892
format version: 4
create_time: 2017-11-09 15:14:49
source replica: mych12
block_id:
merge
20171109_20171109_35414_35419_1
20171109_20171109_35420_35420_0
20171109_20171109_35421_35421_0
20171109_20171109_35422_35422_0
20171109_20171109_35423_35423_0
20171109_20171109_35424_35424_0
20171109_20171109_35425_35425_0
into
20171109_20171109_35414_35425_2
deduplicate: 0

cZxid = 0x14038285a9
ctime = Thu Nov 09 15:14:49 UTC 2017
mZxid = 0x14038285a9
mtime = Thu Nov 09 15:14:49 UTC 2017
pZxid = 0x14038285a9
cversion = 0
dataVersion = 0
aclVersion = 0
ephemeralOwner = 0x0
dataLength = 367
numChildren = 0
```

It just sits in zookeeper and doesn't go anywhere.

Upgrade path for this shard (3 nodes in it) was like this:

1. Everything is running `1.1.54304`.
2. One node (not the problematic one) is downgraded to `1.1.54290`.
3. Everything is upgraded to `1.1.54304`.

**Comment by bobrik** [comment]
I removed problematic queue item and restarted clickhouse, no more errors in logs.

**Comment by alexey-milovidov** [comment]
This is rare issue, when wrong file size or checksum is written in ZooKeeper.

We have the following scenario:
- one replica is doing merge and write merged part to disk;
- simultaneously, checksum and file sizes are calculated and written to ZooKeeper;
- somehow, the data written to disk actually have different checksum or size;
- another replica tries to do the same merge;
- when merge is done, it happens, that data checksum doesn't match the "true" checksum recorded in ZooKeeper;
- it write the message in log and then tries to download already merged part from original replica instead of doing merge locally;
- when the part is downloaded, it happens again, that checksum doesn't match the "true" checksum recorded in ZooKeeper;
- after a while, it will retry, and it happens every time.

We have faced this issue just once. Wrong data was written to disk due to race condition. It was fixed in 1.1.54236. Other reasons are also possible.

> One node (not the problematic one) is downgraded to 1.1.54290.

Please keep in mind that version `1.1.54290` was never being stable!
```
$ git tag --list | grep '1.1.54290'
v1.1.54290-testing
```

Testing versions should go only for testing environments.
Sometimes we find and fix critical issues only after deploying to testing environment.

> I removed problematic queue item and restarted clickhouse, no more errors in logs.

This is correct action.

**Comment by bobrik** [comment]
Thanks, that makes sense. I'll update the ticket if we ever see this again.
